Fixes #33434 - Unify fact importers to one importer

domitea opened the pull request:
This PR is another step of making fact parsers more clear and it's focusing in unify of fact importers. So that means that should be there only one fact importer for importing facts into Foreman. Since it's a bigger change that previous FactParserRegistry (https://github.com/theforeman/foreman/pull/8762) this PR is a draft. It's also lacking tests and they're will be updated but I would like to get some comments of my changes that I made. 

Since the code of importers were usually used as a definition for correct `FactName` class and Smart proxy features, I've moved these code to the respective parsers classes. I think it's better to have them there. 

The little problem was a normalizing of facts. Katello's importer has own normalizing of facts that's only changes separator and normalizes the 'usual' way (by using `StructuredFactImporter`). So I made a set of 'Normalizer' classes that's makes all the normalizing work for importers. The code is cleaner and focuses only to one thing. Also I moved importers under `FactImporters` module to have better organized. I have same intention with the parsers in future. 

Every comment is welcomed!

ekohl commented on app/services/fact_importers/normalizers/katello.rb:
```diff
@@ -0,0 +1,19 @@
+      end
+
+      def change_separator(facts)
+        to_ret = {}
+        facts.each do |key, value|
+          to_ret[key.split('.').join(RhsmFactName::SEPARATOR)] = value
+        end
+        to_ret
```
I always wonder about this, but is this better?
```suggestion
        facts.map do |key, value|
          [key.split('.').join(RhsmFactName::SEPARATOR), value]
        end.to_h
```

Or `Hash[facts.map { ... }]`

Also, split and join; should we instead use `key.gsub('.', RhsmFactName::SEPARATOR)`

ekohl commented on app/services/fact_importers/normalizers/katello.rb:
```diff
@@ -0,0 +1,19 @@
+      def change_separator(facts)
+        to_ret = {}
+        facts.each do |key, value|
+          to_ret[key.split('.').join(RhsmFactName::SEPARATOR)] = value
```
Should this be
```suggestion
          to_ret[key.split('.').join(FactName::SEPARATOR)] = value
```

ekohl commented on app/services/puppet_fact_parser.rb:
```diff
@@ -145,6 +145,14 @@ def disks_total
   end
 
+  def self.smart_proxy_features
+    'Puppet'.freeze
```
I don't think this is correct and that the Puppet feature on the Smart Proxy doesn't have anything to do with fact parsing. It is only an abstraction over Puppetserver's API (and translates the authentication) but all communication from Puppetserver back to Foreman happens directly via the API.

On the other hand, is this used to check if the Smart Proxy is authorized to send facts? If so, I'd try to include that in the method name to make it clear.

domitea replied:
Well, my idea behind moving this into parser is the fact that when I have a parser, I can accept that kind of fact. Also it was originally at importer and this is only parser specific place.

ekohl reviewed (commented):
I had a quick look and this isn't too in-depth. Is a first pass from your side? I hope to remove all database accesses from the actual fact parsers. IMHO only the importer should do so. However, I'd understand it if you did that in a future PR. So in short: what do you consider the scope of this PR?

domitea commented:
> I had a quick look and this isn't too in-depth. Is a first pass from your side? I hope to remove all database accesses from the actual fact parsers. IMHO only the importer should do so. However, I'd understand it if you did that in a future PR. So in short: what do you consider the scope of this PR?

Yes, it's a first pass. Main idea was get rid of plugin specific importers and have only generic one. Thanks to that, I can update the importer logic at one place. Parsers will be in the next PRs. I would like avoid to have big fat PRs with a lot of code.

ekohl commented:
That sounds like a great first pass at unification.

domitea committed:
Fixes #33434 - Unify fact importers

lzap commented:
Let me review the current class responsibilities:

* FactParser - responsible for taking facts and providing interface to recognize host-associated objects like architecture, operating system, domain, hostgroup, cores, ram, total disk size and others as well as interfaces.
* FactImporter - responsible for storing facts into database as well as excluding facts according to excluded_facts setting.

These two classes do different jobs and I think we should keep them separate for the future. More than that, both these classes appear to be God-object antipattern; it tries to do too much. We should break them into smaller classes, perhaps each responsible for a single operation (e.g. recognize OS, recognize RAM or store facts into DB). But I feel like there should be two groups of classes in the pipeline:

* Parser - recognizes Foreman associations or common facts
* Importer - stores data in DB

There was a lengthy post on our Discourse about this, I dropped some of my thoughts there for you to evaluate: https://community.theforeman.org/t/refactoring-of-fact-parsers/23760/11

Now, I believe this PR goes against the goal. If I understand this correctly, you propose offload some of the storing responsibilities from Importer to Parser. That is exactly what Ewoud mentions in comments as well.

Today, both parsing and importing is done in a class called HostFactImporter where facts gets imported first (stored in DB), then facters are called from the registry and parser object is passed into the host object to populate fields. This is a weird design to start with.

I think we should be aiming for a clean pipeline when facts are parsed, then stored in the DB and host is populated. Parsing should be done by parser classes, storing and host populating by importer classes (I refer to them as Serializers in the discourse thread). This design enables plugin authors to write parsers that can affect what will be stored in the DB or host object.

domitea commented:
I understand what do you mean @lzap and as you mentioned in your comment, there should be only one importer. This PR is not aimed to destroy all God-like classes. It's just another step in refactoring. Many more things that you wrote I have planned for future PRs. Reviewers really doesn't like an big PRs with a lot of changes (like 100 files changed) So I try to make relatively small PRs that can be handled.

I don't think that I offload some storing capabilities from `Importer` to `Parsers`. I just moved two methods from plugin-dependent importer to respective parser. First one is the method that only returns correct class for `FactName` and I think that Parser is much better suited for this method because it shows what importer can expect for the Fact. Second one is for `smart_proxy_features`. Ideally, there could be sort of registry explicitly for that (due to plugins) or just array of values on the right places (since the most of parsers are in core). 

Also, I really would like to make changes at the `FactImporters` like you but firstly I need to stack all importers to one because I don't want to change things on many places.

lzap commented:
So what is the status? Sorry I left this uncommented too long.

Bigger refactorings are much easier to review when you keep broken them down to individual commits like "renamed this class", "moved this file here" and similar. Now, I do not want to break this that would be too much work, but can you please give me an in-depth overview of the patch itself? I mean, what you moved where and why.

domitea commented:
@lzap as you're mention in comments before, there is two main components for fact parsing. `FactParser` and `FactImporter`. As we're discussed it earlier, the best option will be when the `FactImporter` will only store data into DB from facts or make another DB related stuff regarding to the parsing like find correct instance of operating system based on the information from `FactParser`.

Problem with these components is that they are plugin specific. That means current codebase has different `FactImporter` for Puppet, Ansible etc. If you're look at the code of these `FactImporters`, you see that they are almost same because they are inherited from `StructuredFactImporter`. So logical step was unify them into one `FactImporter`. They were some leftover methods, so I moved them into respective `FactParsers`, namely `fact_name_class` and `smart_proxy_features`. I really to have some sort of registries (Like `FactParsersRegistry`) but it's a question if it's worth it. 

Also, `FactImporter` is also used as class for normalizing of facts, when they are stored from DB. I'm huge fan of Unix philosophy and especially that "The program should do only one thing and do it well". So I introduced `Normalizer` class, that serve only one thing - To normalize facts for proper store in DB.

Note that there is also `HostFactImporter`. This importer has a purpose to offload some logic from host model to not to be too complex, however from my point of view, it looks like part of importing is handled by `HostFactImporter` and another part by `StructuredFactImporter`. So I move most of DB related import into `StructuredFactImporter`.

I hope that this overview helps @lzap

lzap commented:
Thanks! I wish this PR was done in multiple commits, I see some pure refactoring (moving things around) and some changes. Since this does not have any tests and it is not sure this is working properly, I would rather start fresh with smaller steps.

If anyone wants to pick it up from here, feel free to do it. Thanks!
